# Scrub Data
*“garbage in, garbage out”* 

**Clean dataset step by step:**
1. Change names by team agreement
2. Do we have duplicates?
3. Is this data set contain missing values (Null/NaN) or any wrong values?
4. Drop column "customerID"

In [1]:
# import important libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# use custom function
%run -i 'py/dataframecheck.py'

In [2]:
# read dataset
df = pd.read_csv("Data/WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.head(3)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   

      PaymentMethod MonthlyCharges  TotalCharges Churn  
0  Electronic check          29.85         29.85    No  
1      Mailed check          56.95        1889.5    No  
2      Mailed check          53.85        108.15   Yes  

[3 rows x 21 columns]

In [3]:
# if we want to see all columns, we set this parametr on
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [4]:
df.head(3)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService     MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling     PaymentMethod  MonthlyCharges TotalCharges Churn
0  7590-VHVEG  Female              0     Yes         No       1           No  No phone service             DSL             No          Yes               No          No          No              No  Month-to-month              Yes  Electronic check           29.85        29.85    No
1  5575-GNVDE    Male              0      No         No      34          Yes                No             DSL            Yes           No              Yes          No          No              No        One year               No      Mailed check           56.95       1889.5    No
2  3668-QPYBK    Male              0      No         No       2          Yes                No             DSL            Yes          Yes               No          No          No              No  Month-to-month              Yes      Mailed check           53.85       108.15   Yes

Use custom function to closer veiw to dataset

In [3]:
sum_info(df)

Number of rows: 7043, number of columns: 21


Name   dtypes  Missing  Uniques         First Row                   Last Row
0         customerID   object        0     7043        7590-VHVEG                 3186-AJIEK
1             gender   object        0        2            Female                       Male
2      SeniorCitizen    int64        0        2                 0                          0
3            Partner   object        0        2               Yes                         No
4         Dependents   object        0        2                No                         No
5             tenure    int64        0       73                 1                         66
6       PhoneService   object        0        2                No                        Yes
7      MultipleLines   object        0        3  No phone service                         No
8    InternetService   object        0        3               DSL                Fiber optic
9     OnlineSecurity   object        0        3                No                        Yes
10      OnlineBackup   object        0        3               Yes                         No
11  DeviceProtection   object        0        3                No                        Yes
12       TechSupport   object        0        3                No                        Yes
13       StreamingTV   object        0        3                No                        Yes
14   StreamingMovies   object        0        3                No                        Yes
15          Contract   object        0        3    Month-to-month                   Two year
16  PaperlessBilling   object        0        2               Yes                        Yes
17     PaymentMethod   object        0        4  Electronic check  Bank transfer (automatic)
18    MonthlyCharges  float64        0     1585             29.85                     105.65
19      TotalCharges   object        0     6531             29.85                     6844.5
20             Churn   object        0        2                No                         No

In [6]:
df.head() # show first 5 rows

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService     MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges Churn
0  7590-VHVEG  Female              0     Yes         No       1           No  No phone service             DSL             No          Yes               No          No          No              No  Month-to-month              Yes           Electronic check           29.85        29.85    No
1  5575-GNVDE    Male              0      No         No      34          Yes                No             DSL            Yes           No              Yes          No          No              No        One year               No               Mailed check           56.95       1889.5    No
2  3668-QPYBK    Male              0      No         No       2          Yes                No             DSL            Yes          Yes               No          No          No              No  Month-to-month              Yes               Mailed check           53.85       108.15   Yes
3  7795-CFOCW    Male              0      No         No      45           No  No phone service             DSL            Yes           No              Yes         Yes          No              No        One year               No  Bank transfer (automatic)           42.30      1840.75    No
4  9237-HQITU  Female              0      No         No       2          Yes                No     Fiber optic             No           No               No          No          No              No  Month-to-month              Yes           Electronic check           70.70       151.65   Yes

In [7]:
df.tail() # last 5 rows

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService     MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges Churn
7038  6840-RESVB    Male              0     Yes        Yes      24          Yes               Yes             DSL            Yes           No              Yes         Yes         Yes             Yes        One year              Yes               Mailed check           84.80       1990.5    No
7039  2234-XADUH  Female              0     Yes        Yes      72          Yes               Yes     Fiber optic             No          Yes              Yes          No         Yes             Yes        One year              Yes    Credit card (automatic)          103.20       7362.9    No
7040  4801-JZAZL  Female              0     Yes        Yes      11           No  No phone service             DSL            Yes           No               No          No          No              No  Month-to-month              Yes           Electronic check           29.60       346.45    No
7041  8361-LTMKD    Male              1     Yes         No       4          Yes               Yes     Fiber optic             No           No               No          No          No              No  Month-to-month              Yes               Mailed check           74.40        306.6   Yes
7042  3186-AJIEK    Male              0      No         No      66          Yes                No     Fiber optic            Yes           No              Yes         Yes         Yes             Yes        Two year              Yes  Bank transfer (automatic)          105.65       6844.5    No

In [9]:
df.columns # return list with all column names

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'], dtype='object')

In [10]:
df.shape # return number of rows and columns

(7043, 21)

In [11]:
df.info() # return info about dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
customerID          7043 non-null object
gender              7043 non-null object
SeniorCitizen       7043 non-null int64
Partner             7043 non-null object
Dependents          7043 non-null object
tenure              7043 non-null int64
PhoneService        7043 non-null object
MultipleLines       7043 non-null object
InternetService     7043 non-null object
OnlineSecurity      7043 non-null object
OnlineBackup        7043 non-null object
DeviceProtection    7043 non-null object
TechSupport         7043 non-null object
StreamingTV         7043 non-null object
StreamingMovies     7043 non-null object
Contract            7043 non-null object
PaperlessBilling    7043 non-null object
PaymentMethod       7043 non-null object
MonthlyCharges      7043 non-null float64
TotalCharges        7043 non-null object
Churn               7043 non-null object
dtypes: float64(1), int64(2), obj

In [13]:
df.head() # shows first 5 rows
df.tail() # shows 5 rows
df.columns # return list with all column names
df.shape # return number of rows and columns
df.info() # return info about dataframe

# count unique values in a column
for col in df.columns:
    df[col].value_counts(dropna=False) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
customerID          7043 non-null object
gender              7043 non-null object
SeniorCitizen       7043 non-null int64
Partner             7043 non-null object
Dependents          7043 non-null object
tenure              7043 non-null int64
PhoneService        7043 non-null object
MultipleLines       7043 non-null object
InternetService     7043 non-null object
OnlineSecurity      7043 non-null object
OnlineBackup        7043 non-null object
DeviceProtection    7043 non-null object
TechSupport         7043 non-null object
StreamingTV         7043 non-null object
StreamingMovies     7043 non-null object
Contract            7043 non-null object
PaperlessBilling    7043 non-null object
PaymentMethod       7043 non-null object
MonthlyCharges      7043 non-null float64
TotalCharges        7043 non-null object
Churn               7043 non-null object
dtypes: float64(1), int64(2), obj

In [16]:
# count unique values in a column
for col in df.columns:
    print(df[col].value_counts(dropna=False))

8095-WANWK    1
6615-ZGEDR    1
1291-CUOCY    1
5686-CMAWK    1
5126-RCXYW    1
             ..
6215-NQCPY    1
8708-XPXHZ    1
4718-WXBGI    1
6369-MCAKO    1
3786-WOVKF    1
Name: customerID, Length: 7043, dtype: int64
Male      3555
Female    3488
Name: gender, dtype: int64
0    5901
1    1142
Name: SeniorCitizen, dtype: int64
No     3641
Yes    3402
Name: Partner, dtype: int64
No     4933
Yes    2110
Name: Dependents, dtype: int64
1     613
72    362
2     238
3     200
4     176
71    170
5     133
7     131
8     123
70    119
9     119
12    117
10    116
6     110
13    109
68    100
15     99
11     99
67     98
18     97
69     95
24     94
22     90
66     89
35     88
17     87
23     85
64     80
56     80
16     80
52     80
25     79
26     79
14     76
65     76
61     76
60     76
46     74
19     73
63     72
29     72
27     72
30     72
20     71
62     70
41     70
53     70
32     69
54     68
47     68
50     68
51     68
58     67
49     66
43     65
42     65
3

In [4]:
inf = sum_info(df)

Number of rows: 7043, number of columns: 21


In [42]:
def sum_info(df):
    print("Number of rows: {0}, number of columns: {1}".format(df.shape[0], df.shape[1]))
    info = pd.DataFrame(df.dtypes, columns=['dtypes'])
    info = info.reset_index()
    info['Name'] = info['index']
    info = info[['Name', 'dtypes']]
    info['Uniques'] = df.nunique().values
    for i in range(0,2):
        info['Row N'+str(i+1)] = df.loc[i].values
    info['Last Row'] = df.loc[df.shape[0]-1].values

    return info

In [43]:
sum_info(df)

Number of rows: 7043, number of columns: 21


Name   dtypes  Uniques            Row N1        Row N2                   Last Row
0         customerID   object     7043        7590-VHVEG    5575-GNVDE                 3186-AJIEK
1             gender   object        2            Female          Male                       Male
2      SeniorCitizen    int64        2                 0             0                          0
3            Partner   object        2               Yes            No                         No
4         Dependents   object        2                No            No                         No
5             tenure    int64       73                 1            34                         66
6       PhoneService   object        2                No           Yes                        Yes
7      MultipleLines   object        3  No phone service            No                         No
8    InternetService   object        3               DSL           DSL                Fiber optic
9     OnlineSecurity   object        3                No           Yes                        Yes
10      OnlineBackup   object        3               Yes            No                         No
11  DeviceProtection   object        3                No           Yes                        Yes
12       TechSupport   object        3                No            No                        Yes
13       StreamingTV   object        3                No            No                        Yes
14   StreamingMovies   object        3                No            No                        Yes
15          Contract   object        3    Month-to-month      One year                   Two year
16  PaperlessBilling   object        2               Yes            No                        Yes
17     PaymentMethod   object        4  Electronic check  Mailed check  Bank transfer (automatic)
18    MonthlyCharges  float64     1585             29.85         56.95                     105.65
19      TotalCharges   object     6531             29.85        1889.5                     6844.5
20             Churn   object        2                No            No                         No

In [24]:
info_table = sum_info(df)

Number of rows: 7043, number of columns: 21


In [25]:
info_table

Name   dtypes  Uniques            Row N1        Row N2                   Last Row
0         customerID   object     7043        7590-VHVEG    5575-GNVDE                 3186-AJIEK
1             gender   object        2            Female          Male                       Male
2      SeniorCitizen    int64        2                 0             0                          0
3            Partner   object        2               Yes            No                         No
4         Dependents   object        2                No            No                         No
5             tenure    int64       73                 1            34                         66
6       PhoneService   object        2                No           Yes                        Yes
7      MultipleLines   object        3  No phone service            No                         No
8    InternetService   object        3               DSL           DSL                Fiber optic
9     OnlineSecurity   object        3                No           Yes                        Yes
10      OnlineBackup   object        3               Yes            No                         No
11  DeviceProtection   object        3                No           Yes                        Yes
12       TechSupport   object        3                No            No                        Yes
13       StreamingTV   object        3                No            No                        Yes
14   StreamingMovies   object        3                No            No                        Yes
15          Contract   object        3    Month-to-month      One year                   Two year
16  PaperlessBilling   object        2               Yes            No                        Yes
17     PaymentMethod   object        4  Electronic check  Mailed check  Bank transfer (automatic)
18    MonthlyCharges  float64     1585             29.85         56.95                     105.65
19      TotalCharges   object     6531             29.85        1889.5                     6844.5
20             Churn   object        2                No            No                         No

In [5]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

### What we can see here?
* We have one entry for each CustomerId
* Dataset is pretty clean, doesn't have missing values
* Some features are categorical, some a numeric

## 1. Change names

For the convenience of working with data, we change all capital letters to lowercase

In [4]:
# for each column set all letters to lower case
df.columns = map(str.lower, df.columns)

In [5]:
df.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice     multiplelines internetservice onlinesecurity onlinebackup deviceprotection techsupport streamingtv streamingmovies        contract paperlessbilling              paymentmethod  monthlycharges totalcharges churn
0  7590-VHVEG  Female              0     Yes         No       1           No  No phone service             DSL             No          Yes               No          No          No              No  Month-to-month              Yes           Electronic check           29.85        29.85    No
1  5575-GNVDE    Male              0      No         No      34          Yes                No             DSL            Yes           No              Yes          No          No              No        One year               No               Mailed check           56.95       1889.5    No
2  3668-QPYBK    Male              0      No         No       2          Yes                No             DSL            Yes          Yes               No          No          No              No  Month-to-month              Yes               Mailed check           53.85       108.15   Yes
3  7795-CFOCW    Male              0      No         No      45           No  No phone service             DSL            Yes           No              Yes         Yes          No              No        One year               No  Bank transfer (automatic)           42.30      1840.75    No
4  9237-HQITU  Female              0      No         No       2          Yes                No     Fiber optic             No           No               No          No          No              No  Month-to-month              Yes           Electronic check           70.70       151.65   Yes

In [6]:
# select columns name for future work
list(df.columns.values.tolist()) 

['customerid',
 'gender',
 'seniorcitizen',
 'partner',
 'dependents',
 'tenure',
 'phoneservice',
 'multiplelines',
 'internetservice',
 'onlinesecurity',
 'onlinebackup',
 'deviceprotection',
 'techsupport',
 'streamingtv',
 'streamingmovies',
 'contract',
 'paperlessbilling',
 'paymentmethod',
 'monthlycharges',
 'totalcharges',
 'churn']

## 2. Check dublicates
**customerID** is column with all uniques values, how we see before in table, but make a double check  

In [7]:
df.duplicated().any()

False

## 3. Missing values (Null/NaN)

How we see before this data set is clean, and doesn't have any Null or Nan in the rows.

In [8]:
df.isnull().sum(axis = 0)

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

Let's check dataset for valid information into rows

In [9]:
for col in df.columns:
    df1 = df[col].value_counts().sort_values(axis=0)
    print(col,"\n", df1 , "\n\n") 

customerid 
 9499-XPZXM    1
9732-KPKBW    1
5895-QSXOD    1
3509-GWQGF    1
5797-APWZC    1
             ..
9693-XMUOB    1
8722-NGNBH    1
3590-TCXTB    1
1031-IIDEO    1
6352-GIGGQ    1
Name: customerid, Length: 7043, dtype: int64 


gender 
 Female    3488
Male      3555
Name: gender, dtype: int64 


seniorcitizen 
 1    1142
0    5901
Name: seniorcitizen, dtype: int64 


partner 
 Yes    3402
No     3641
Name: partner, dtype: int64 


dependents 
 Yes    2110
No     4933
Name: dependents, dtype: int64 


tenure 
 0      11
36     50
44     51
39     56
28     57
38     59
59     60
45     61
21     63
33     64
48     64
40     64
55     64
42     65
31     65
43     65
37     65
34     65
57     65
49     66
58     67
51     68
50     68
47     68
54     68
32     69
41     70
62     70
53     70
20     71
30     72
27     72
29     72
63     72
19     73
46     74
60     76
65     76
61     76
14     76
26     79
25     79
52     80
16     80
56     80
64     80
23     85
17    

In closer look **TotalCharges** should has type float64, but it is object. We can see empty values, let's remove them

In [10]:
df['totalcharges'] = df["totalcharges"].replace(" ",np.nan) 
df['totalcharges'].isna().sum()

11

In [11]:
# count the percentage of missing data
round(df['totalcharges'].isna().sum()/len(df)*100,5)

0.15618

Percentage of empty values is small, we can drop this rows from dataset lossless.<br>
But let's check it one more time, may be we can calculate this column == tenure * monthlycharges

In [12]:
df.loc[df["totalcharges"].isna()]

customerid  gender  seniorcitizen partner dependents  tenure phoneservice     multiplelines internetservice       onlinesecurity         onlinebackup     deviceprotection          techsupport          streamingtv      streamingmovies  contract paperlessbilling              paymentmethod  monthlycharges totalcharges churn
488   4472-LVYGI  Female              0     Yes        Yes       0           No  No phone service             DSL                  Yes                   No                  Yes                  Yes                  Yes                   No  Two year              Yes  Bank transfer (automatic)           52.55          NaN    No
753   3115-CZMZD    Male              0      No        Yes       0          Yes                No              No  No internet service  No internet service  No internet service  No internet service  No internet service  No internet service  Two year               No               Mailed check           20.25          NaN    No
936   5709-LVOEQ  Female              0     Yes        Yes       0          Yes                No             DSL                  Yes                  Yes                  Yes                   No                  Yes                  Yes  Two year               No               Mailed check           80.85          NaN    No
1082  4367-NUYAO    Male              0     Yes        Yes       0          Yes               Yes              No  No internet service  No internet service  No internet service  No internet service  No internet service  No internet service  Two year               No               Mailed check           25.75          NaN    No
1340  1371-DWPAZ  Female              0     Yes        Yes       0           No  No phone service             DSL                  Yes                  Yes                  Yes                  Yes                  Yes                   No  Two year               No    Credit card (automatic)           56.05          NaN    No
3331  7644-OMVMY    Male              0     Yes        Yes       0          Yes                No              No  No internet service  No internet service  No internet service  No internet service  No internet service  No internet service  Two year               No               Mailed check           19.85          NaN    No
3826  3213-VVOLG    Male              0     Yes        Yes       0          Yes               Yes              No  No internet service  No internet service  No internet service  No internet service  No internet service  No internet service  Two year               No               Mailed check           25.35          NaN    No
4380  2520-SGTTA  Female              0     Yes        Yes       0          Yes                No              No  No internet service  No internet service  No internet service  No internet service  No internet service  No internet service  Two year               No               Mailed check           20.00          NaN    No
5218  2923-ARZLG    Male              0     Yes        Yes       0          Yes                No              No  No internet service  No internet service  No internet service  No internet service  No internet service  No internet service  One year              Yes               Mailed check           19.70          NaN    No
6670  4075-WKNIU  Female              0     Yes        Yes       0          Yes               Yes             DSL                   No                  Yes                  Yes                  Yes                  Yes                   No  Two year               No               Mailed check           73.35          NaN    No
6754  2775-SEFEE    Male              0      No        Yes       0          Yes               Yes             DSL                  Yes                  Yes                   No                  Yes                   No                   No  Two year              Yes  Bank transfer (automatic)           61.90          NaN    No

In [13]:
# dropping empty values from TotalCharges column which contain .15% missing data 
df.dropna(inplace=True)

In [14]:
#convert to float type
df["totalcharges"] = df["totalcharges"].astype(float)

In [15]:
df.shape

(7032, 21)

## 4. Drop column

In [16]:
# customerid not useful better to drop it
df.drop(['customerid'], axis=1, inplace=True)

In [17]:
df.shape

(7032, 20)

## 5. Phone Service vs. Multiple Lines

The information included in the 'phoneservice' column is in the 'multipleines' column. We need to do more for better anderstanding this column.

## 6. Internet Services

In [18]:
#check columns
internet_cols = [ 'onlinesecurity', 'onlinebackup', 'deviceprotection',
                'techsupport','streamingtv', 'streamingmovies']
for col in internet_cols : 
    print(col,"\n", df[col].value_counts() , "\n\n") 

onlinesecurity 
 No                     3497
Yes                    2015
No internet service    1520
Name: onlinesecurity, dtype: int64 


onlinebackup 
 No                     3087
Yes                    2425
No internet service    1520
Name: onlinebackup, dtype: int64 


deviceprotection 
 No                     3094
Yes                    2418
No internet service    1520
Name: deviceprotection, dtype: int64 


techsupport 
 No                     3472
Yes                    2040
No internet service    1520
Name: techsupport, dtype: int64 


streamingtv 
 No                     2809
Yes                    2703
No internet service    1520
Name: streamingtv, dtype: int64 


streamingmovies 
 No                     2781
Yes                    2731
No internet service    1520
Name: streamingmovies, dtype: int64 




If user doesn't have internet (No internet service) than he can use any of this internet services. Can we replace value "No internet service" to "No"? 

In [19]:
df.shape

(7032, 20)

## Conclusion
Save clean data set into new file:

In [20]:
df.to_csv("data/clean_churn.csv", index=False)

We ready for EDA with data set in 7032 rows and 20 columns.